src/sql/src/plan/statement/dml.rs

```diff
@@ -457,16 +454,52 @@ pub fn plan_subscribe(
                 QueryLifetime::OneShot(scx.pcx()?),
             )?;
             assert!(query.finishing.is_trivial(query.desc.arity()));
-            SubscribeFrom::Query {
-                expr: query.expr,
-                desc: query.desc,
-            }
+            let desc = query.desc.clone();
+            (
+                SubscribeFrom::Query {
+                    expr: query.expr,
+                    desc: query.desc,
+                },
+                desc,
+            )
         }
     };
 
     let when = query::plan_as_of(scx, as_of)?;
     let up_to = up_to.map(|up_to| plan_up_to(scx, up_to)).transpose()?;
 
+    let key_indices = envelope
+        .map(|env| match env {
+            SubscribeEnvelope::Upsert { key_columns } => {
+                scx.require_envelope_upsert_in_subscribe()?;
+                let key_columns = key_columns
+                    .into_iter()
+                    .map(normalize::column_name)
+                    .collect::<Vec<_>>();
+                let mut uniq = BTreeSet::new();
+                for col in key_columns.iter() {
+                    if !uniq.insert(col) {
+                        sql_bail!("Repeated column name in subscribe envelope key: {}", col);
+                    }
+                }
+                let indices = key_columns
+                    .iter()
+                    .map(|col| -> anyhow::Result<usize> {
+                        let name_idx = desc
+                            .get_by_name(col)
+                            .map(|(idx, _type)| idx)
+                            .ok_or_else(|| sql_err!("No such column: {}", col))?;
+                        if desc.get_unambiguous_name(name_idx).is_none() {
+                            anyhow::bail!("Ambiguous column: {}", col);
```

Comment:
Why is this an `anyhow` instead of `sql_err`?

Reply:
Fixed. It was a copy paste